gunpowder/jax/nodes/train.py

```diff
@@ -0,0 +1,288 @@
+            Directory for saving tensorboard summaries.
+
+        log_every (``int``, optional):
+
+            After how many iterations to write out tensorboard summaries.
+
+        spawn_subprocess (``bool``, optional):
+
+            Whether to run the ``train_step`` in a separate process. Default is
+            false.
+
+        n_devices (``int``, optional):
+
+            Number of GPU devices to train on concurrently using `jax.pmap`. If
+            `None`, the number of available GPUs will be automatically detected
+            and used.
+    """
+
+    def __init__(
+        self,
+        model: GenericJaxModel,
+        inputs: Dict[str, ArrayKey],
+        outputs: Dict[Union[int, str], ArrayKey],
+        gradients: Dict[Union[int, str], ArrayKey] = {},
+        array_specs: Optional[Dict[ArrayKey, ArraySpec]] = None,
+        checkpoint_basename: str = "model",
+        save_every: int = 2000,
+        keep_n_checkpoints: Optional[int] = None,
+        log_dir: str = None,
+        log_every: int = 1,
+        spawn_subprocess: bool = False,
+        n_devices: Optional[int] = None
+    ):
+
+        # not yet implemented
+        gradients = gradients
+
+        super(Train, self).__init__(
+            inputs, outputs, gradients, array_specs,
+            spawn_subprocess=spawn_subprocess
+        )
+
+        self.model = model
+        self.checkpoint_basename = checkpoint_basename
+        self.save_every = save_every
+        if n_devices is None:
+            n_devices = jax.local_device_count()  # autodetect available GPUs
+        self.n_devices = n_devices
+        self.local_devices = jax.local_devices()
+        self.keep_n_checkpoints = keep_n_checkpoints
+
+        self.iteration = 0
+
+        if not isinstance(tensorboardX, NoSuchModule) and log_dir is not None:
+            self.summary_writer = tensorboardX.SummaryWriter(log_dir)
+            self.log_every = log_every
+        else:
+            self.summary_writer = None
+            if log_dir is not None:
+                logger.warning("log_dir given, but tensorboardX is not installed")
+
+        self.intermediate_layers = {}
+
+    def replicate_params(self, params):
+        return jax.tree_map(lambda x: jnp.array([x] * self.n_devices), params)
+
+    def start(self):
+        checkpoint, self.iteration = self._get_latest_checkpoint(
+            self.checkpoint_basename)
+
+        if checkpoint is not None:
+
+            logger.info("Resuming training from iteration %d", self.iteration)
+
+            with open(checkpoint, 'rb') as f:
+                self.model_params = pickle.load(f)
+                if self.n_devices > 1:
+                    self.model_params = self.replicate_params(self.model_params)
+        else:
+
+            logger.info("Starting training from scratch")
+            self.model_params = None
+
+    def split_inputs(self, inputs):
+        for k, arr in inputs.items():
+            assert arr.shape[0] % self.n_devices == 0
```

Comment:
This should maybe be documented. I'm guessing shape[0] will usually be the batch dimension. Might be obvious enough that batch size should be a multiple of the number of gpus you have available.
Not super familiar with how this works, can you parallelize computation over any axis? If so might be a weird constraint that some spatial dimension of your input needs to be a multiple of the number of gpus available.

Reply:
For now JAX can only `pmap` across the first dimension. Will add some comment